spark/src/main/scala/geotrellis/spark/costdistance/MrGeoCostDistance.scala

```diff
@@ -0,0 +1,187 @@
+import org.apache.spark.SparkContext
+import org.apache.spark.util.AccumulatorV2
+import org.apache.spark.storage.StorageLevel
+
+import scala.collection.mutable
+
+
+/**
+  * This Spark-enabled implementation of the standard cost-distance
+  * algorithm [1] is "heavily inspired" by the MrGeo implementation
+  * [2] but does not share any code with it.
+  *
+  * 1. Tomlin, Dana.
+  *    "Propagating radial waves of travel cost in a grid."
+  *    International Journal of Geographical Information Science 24.9 (2010): 1391-1413.
+  *
+  * 2. https://github.com/ngageoint/mrgeo/blob/0c6ed4a7e66bb0923ec5c570b102862aee9e885e/mrgeo-mapalgebra/mrgeo-mapalgebra-costdistance/src/main/scala/org/mrgeo/mapalgebra/CostDistanceMapOp.scala
+  */
+object MrGeoCostDistance {
+
+  type CostList = List[CostDistance.Cost]
+  type KeyListPair = (SpatialKey, CostList)
+  type Changes = mutable.Map[SpatialKey, CostList]
+
+  /**
+    * An accumulator to hold lists of edge changes.
+    */
+  class ChangesAccumulator extends AccumulatorV2[KeyListPair, Changes] {
+    private var map: Changes = mutable.Map.empty
+
+    def copy: ChangesAccumulator = {
+      val other = new ChangesAccumulator
+      other.merge(this)
+      other
+    }
+    def add(kv: (SpatialKey, CostList)): Unit = {
+      val key = kv._1
+      val list = kv._2
+
+      if (!map.contains(key)) map.put(key, list)
+      else map.put(key, map.get(key).get ++ list)
+    }
+    def isZero: Boolean = map.isEmpty
+    def merge(other: AccumulatorV2[(SpatialKey, CostList), Changes]): Unit = map ++= other.value
+    def reset: Unit = map.clear
+    def value: Changes = map
+  }
+
+  /**
+    * Perform the cost-distance computation.
+    */
+  def apply[K: (? => SpatialKey), V: (? => Tile)](
+    friction: RDD[(K, V)] with Metadata[TileLayerMetadata[K]],
+    points: List[Point],
+    maxCost: Double = Double.PositiveInfinity
+  )(implicit sc: SparkContext): RDD[(K, Tile)] = {
+
+    val mt = friction.metadata.mapTransform
+    val bounds = friction.metadata.bounds.asInstanceOf[KeyBounds[K]]
+    val minKey = implicitly[SpatialKey](bounds.minKey)
+    val maxKey = implicitly[SpatialKey](bounds.maxKey)
+
+    val accumulator = new ChangesAccumulator
+    sc.register(accumulator, "Changes")
+
+    // Load the accumulator with the starting values
+    friction.foreach({ case (k, v) =>
+      val key = implicitly[SpatialKey](k)
+      val tile = implicitly[Tile](v)
+      val cols = tile.cols
+      val rows = tile.rows
+      val extent = mt(key)
+      val rasterExtent = RasterExtent(extent, cols, rows)
+      val costs: List[CostDistance.Cost] = points
+        .filter({ point => extent.contains(point) })
+        .map({ point =>
+          val col = rasterExtent.mapXToGrid(point.x)
+          val row = rasterExtent.mapYToGrid(point.y)
+          val friction = tile.getDouble(col, row)
+          (col, row, friction, 0.0)
+        })
+      accumulator.add((key, costs))
+    })
```

Comment:
I see the loop version does a `count` to force the execution and cache the RDD. Is it possible to avoid this `foreach` with a similar method, where this accumulator is moved to the `map`, that RDD persisted, and the RDD counted to force execution? Perhaps not the best optimization, but your opinion on this would probably help me understand the logic better.